test/SILOptimizer/functionsigopts.sil

```diff
@@ -1553,3 +1775,15 @@ bb0(%0 : $Builtin.Int32):
 // CHECK:  tuple
 // CHECK:  return
 
+// Check that a owned-to-guaranteed has happened.
+// CHECK-LABEL: sil hidden [noinline] @_T027generic_owned_to_guaranteedTf4g_n : $@convention(thin) <T where T : KlassFoo> (@guaranteed T) -> Int64
```

Comment:
Can you also check the body of the specialized functions? E.g. check that the destroy_value was eliminated?

Otherwise LGTM!